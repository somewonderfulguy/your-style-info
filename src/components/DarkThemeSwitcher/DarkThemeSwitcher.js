import React from 'react'

const DarkThemeSwitcher = () => (
  <div>
    Switch
  </div>
)

export default DarkThemeSwitcher