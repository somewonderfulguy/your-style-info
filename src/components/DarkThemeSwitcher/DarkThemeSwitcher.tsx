import { useThemeState, useThemeSwitch } from '~contexts/themeContext'

// TODO: colocate images with component
import moonIcon from '~assets/images/moon.svg'
import sunIcon from '~assets/images/sun.svg'
import styles from './DarkThemeSwitcher.module.css'

type Props = {
  darkerPalette?: boolean
  labelText?: boolean
}

const DarkThemeSwitcher = ({
  darkerPalette = false,
  labelText = false
}: Props) => {
  const isDarkTheme = useThemeState()
  const switchTheme = useThemeSwitch()

  console.log(moonIcon)

  return (
    // TODO: dynamically change aria label text
    <label
      className={darkerPalette ? styles.switchDarker : styles.switch}
      aria-label="switch theme"
    >
      <input
        className={styles.checkbox}
        type="checkbox"
        checked={isDarkTheme}
        onChange={switchTheme}
      />
      <div className={styles.slider}>
        <div
          className={styles.round}
          style={{
            backgroundImage: `url("${sunIcon.src}")`,
            opacity: isDarkTheme ? 0 : 1
          }}
        />
        <div
          className={styles.round}
          style={{
            backgroundImage: `url("${moonIcon.src}")`,
            opacity: isDarkTheme ? 1 : 0,
            zIndex: 1
          }}
        />
      </div>
      {labelText && <span className={styles.text}>Switch theme</span>}
    </label>
  )
}

export default DarkThemeSwitcher
