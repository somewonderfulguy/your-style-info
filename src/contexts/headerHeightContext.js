import React, {createContext, useContext, useMemo, useState} from 'react'

const HeaderHeightContext = createContext()

const useHeaderHeight = () => {
  const context = useContext(HeaderHeightContext)
  if(!context) {
    throw new Error('useHeaderHeight must be used within a HeaderHeightProvider')
  }
  return context
}

const HeaderHeightProvider = props => {
  const [headerHeight, setHeaderHeight] = useState(0)
  const value = useMemo(() => ({headerHeight, setHeaderHeight}), [headerHeight])
  return <HeaderHeightContext.Provider value={value} {...props} />
}

export {HeaderHeightProvider, useHeaderHeight}