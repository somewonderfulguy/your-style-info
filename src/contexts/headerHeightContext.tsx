import {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useContext,
  useState
} from 'react'

const HeaderHeightStateContext = createContext<number | undefined>(undefined)
HeaderHeightStateContext.displayName = 'HeaderHeightStateContext'

const HeaderHeightDispatchContext = createContext<
  Dispatch<SetStateAction<number>> | undefined
>(undefined)
HeaderHeightDispatchContext.displayName = 'HeaderHeightDispatchContext'

const HeaderHeightProvider = ({ children }: { children: ReactNode }) => {
  const [headerHeight, setHeaderHeight] = useState(0)
  return (
    <HeaderHeightStateContext.Provider value={headerHeight}>
      <HeaderHeightDispatchContext.Provider value={setHeaderHeight}>
        {children}
      </HeaderHeightDispatchContext.Provider>
    </HeaderHeightStateContext.Provider>
  )
}

const useHeaderHeightState = () => {
  const context = useContext(HeaderHeightStateContext)
  if (context === undefined) {
    throw new Error(
      'useHeaderHeightState must be used within a HeaderHeightProvider'
    )
  }
  return context
}

const useHeaderHeightDispatch = () => {
  const context = useContext(HeaderHeightDispatchContext)
  if (context === undefined) {
    throw new Error(
      'useHeaderHeightDispatch must be used within a HeaderHeightProvider'
    )
  }
  return context
}

export { HeaderHeightProvider, useHeaderHeightState, useHeaderHeightDispatch }
