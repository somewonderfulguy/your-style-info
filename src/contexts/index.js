export * from './headerHeightContext'
export * from './loadingContext'
export * from './localisationContext'
export * from './screenDimensionsContext'
export * from './themeContext'