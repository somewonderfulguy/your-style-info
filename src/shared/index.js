export * from './componentRenderer'
export * from './getComponent'
export * from './tests'