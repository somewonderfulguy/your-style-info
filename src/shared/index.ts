export * from './hooks'
export * from './tests'
export * from './utils'

export * from './types'