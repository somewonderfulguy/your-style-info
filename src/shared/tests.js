import React from 'react'
import {render as rtlRender} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {Router} from 'react-router-dom'
import {createMemoryHistory} from 'history'
import {QueryClient, QueryClientProvider} from 'react-query'

import {HeaderHeightProvider, LocalizationProvider, ScreenDimensionsProvider, ThemeProvider} from 'contexts'

const queryClient = new QueryClient({
  defaultOptions: {
    initialStale: true
  }
})

const render = (ui, {route = '/', ...options} = {}) => {
  const history = createMemoryHistory({initialEntries: [route]})
  const Wrapper = props => (
    <QueryClientProvider client={queryClient}>
      <Router history={history}>
        <HeaderHeightProvider>
          <LocalizationProvider>
            <ThemeProvider>
              <ScreenDimensionsProvider {...props} />
            </ThemeProvider>
          </LocalizationProvider>
        </HeaderHeightProvider>
      </Router>
    </QueryClientProvider>
  )
  return {
    ...rtlRender(ui, {wrapper: Wrapper, ...options}),
    history,
  }
}

const renderWithoutProviders = rtlRender

export * from '@testing-library/react'
export {render, renderWithoutProviders, userEvent}