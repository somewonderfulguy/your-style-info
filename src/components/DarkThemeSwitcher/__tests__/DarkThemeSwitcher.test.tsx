import { renderHook } from '@testing-library/react-hooks'
import user from '@testing-library/user-event'

import { screen, render, renderWholeApp } from 'shared/tests'
import { ThemeProvider, useThemeState } from 'contexts'
import DarkThemeSwitcher from '..'

const setup = (darkerPalette = false, labelText = false) => {
  const utils = render(
    <DarkThemeSwitcher darkerPalette={darkerPalette} labelText={labelText} />
  )
  return utils
}

test('should match snapshot', () => {
  const { asFragment } = setup()
  expect(asFragment()).toMatchSnapshot()
})

test('should change depending on darkerPalette prop', () => {
  const defaultProps = setup().asFragment()
  const withDarkPalette = setup(true).asFragment()

  expect(defaultProps).toMatchDiffSnapshot(withDarkPalette, {
    contextLines: 1,
    aAnnotation: 'dark palette off',
    bAnnotation: 'dark palette on'
  })
})

test('should change depending on labelText prop', () => {
  const defaultProps = setup().asFragment()
  const withLabelText = setup(false, true).asFragment()

  expect(defaultProps).toMatchDiffSnapshot(withLabelText, {
    contextLines: 1,
    aAnnotation: 'label text off',
    bAnnotation: 'label text on'
  })
})

test('switching theme should work properly', async () => {
  renderWholeApp()

  const [themeSwitcher] = screen.getAllByLabelText(/switch theme/i)

  const getHook = () =>
    renderHook(() => useThemeState(), {
      wrapper: ThemeProvider
    })

  // default (light)
  expect(getHook().result.current).toBeFalsy()
  expect(window.localStorage.getItem('isDarkTheme')).toBe('false')

  // switch to dark
  user.click(themeSwitcher)
  expect(getHook().result.current).toBeTruthy()
  expect(window.localStorage.getItem('isDarkTheme')).toBe('true')

  // switch to light
  user.click(themeSwitcher)
  expect(getHook().result.current).toBeFalsy()
  expect(window.localStorage.getItem('isDarkTheme')).toBe('false')
})
