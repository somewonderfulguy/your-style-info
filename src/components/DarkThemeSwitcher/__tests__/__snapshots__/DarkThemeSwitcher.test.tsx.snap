// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change depending on darkerPalette prop 1`] = `
Snapshot Diff:
- dark palette off
+ dark palette on

@@ -3,3 +3,3 @@
      aria-label="switch theme"
-     class="switch"
+     class="switchDarker"
    >
`;

exports[`should change depending on labelText prop 1`] = `
Snapshot Diff:
- label text off
+ label text on

@@ -21,2 +21,7 @@
      </div>
+     <span
+       class="text"
+     >
+       Switch theme
+     </span>
    </label>
`;

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <label
    aria-label="switch theme"
    class="switch"
  >
    <input
      class="checkbox"
      type="checkbox"
    />
    <div
      class="slider"
    >
      <div
        class="round"
        style="background-image: url(sun.svg); opacity: 1;"
      />
      <div
        class="round"
        style="background-image: url(moon.svg); opacity: 0; z-index: 1;"
      />
    </div>
  </label>
</DocumentFragment>
`;
